pkg/config/crd/validator.go

```diff
@@ -0,0 +1,214 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package crd
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/go-openapi/validate"
+	"github.com/hashicorp/go-multierror"
+	"k8s.io/apiextensions-apiserver/pkg/apis/apiextensions"
+	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	apiextensionsv1beta1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1"
+	"k8s.io/apiextensions-apiserver/pkg/apiserver/validation"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	kubeyaml "k8s.io/apimachinery/pkg/util/yaml"
+	"k8s.io/client-go/kubernetes/scheme"
+	serviceapis "sigs.k8s.io/service-apis/apis/v1alpha1"
+
+	clientnetworkingalpha "istio.io/client-go/pkg/apis/networking/v1alpha3"
+	clientnetworkingbeta "istio.io/client-go/pkg/apis/networking/v1beta1"
+	clientsecurity "istio.io/client-go/pkg/apis/security/v1beta1"
+	"istio.io/istio/pkg/test"
+	"istio.io/istio/pkg/test/env"
+	"istio.io/istio/pkg/test/util/yml"
+)
+
+// Validator returns a new validator for custom resources
+// Warning: this is meant for usage in tests only
+type Validator struct {
+	byGvk map[schema.GroupVersionKind]*validate.SchemaValidator
+	// If enabled, resources without a validator will be ignored. Otherwise, they will fail.
+	SkipMissing bool
+	Scheme      *runtime.Scheme
+}
+
+func (v *Validator) ValidateCustomResourceYAML(data string) error {
+	decoder := serializer.NewCodecFactory(v.Scheme, serializer.EnableStrict).UniversalDeserializer()
+
+	var errs *multierror.Error
+	for _, item := range yml.SplitString(data) {
+		obj, _, err := decoder.Decode([]byte(item), nil, nil)
+		if err != nil {
+			return err
+		}
+		errs = multierror.Append(errs, v.ValidateCustomResource(obj))
+	}
+	return errs.ErrorOrNil()
+}
+
+func (v *Validator) ValidateCustomResource(o runtime.Object) error {
```

Comment:
How confident are we with this approach?

How do we plan to use this validator with our config files? Can we setup a pipeline in presubmit that takes config files and dump into a live k8s cluster and validate there?

Reply:
It's not 100% accurate as noted in the workaround part where we filter out some errors. I would hope that at some point we figure out how to do this properly. I do not want our unit tests requiring a live cluster.

This isn't perfect but it gives us strictly more coverage than today do I think it's still useful. I wouldn't expose to users (like integrate with analyze) until we iron out the kinks though

Reply:
How do we ensure coverage on config files?

Reply:
I am not sure what you mean?

Current state: we have a bunch of configs that may be invalid

With this PR: we have a bunch of configs that may be invalid, but for some of them that we call ValidateCustomResource we have some degree of confidence they are valid

Reply:
So the next step is to have all "readconfig" funcs use the validator?

Reply:
yes, if they test wants to enforce the configs passed are valid they should do this. Currently I have added it to one test where its a major issue and constantly falling out of sync with the schema validation

Reply:
Ok. Can we file a bug or add a TODO that we need to provide a more reliable way for validation testing? Something that we can keep track.